oxenmq/pubsub.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include "connections.h"
+
+#include <unordered_map>
+#include <shared_mutex>
+#include <chrono>
+
+namespace oxenmq
+{
+
+using namespace std::chrono_literals;
+
+struct DummySub {
+    bool operator==(DummySub& other) { return true; }
+};
+
+/**
+ * OMQ Subscription class.  Handles pub/sub connections such that the user
+ * only needs to call methods to subscribe and publish.
+ * FIXME: do we want an unsubscribe, or is expiry / conn management sufficient?
+ *
+ * type T can contain whatever information the user may need at publish time,
+ * for example if the subscription is for logs the subscriber can specify log
+ * levels or categories, and the publisher can choose to send or not based on
+ * those.
+ */
+template <typename T = DummySub>
+class Subscription
+{
+    struct SubData
+    {
+        std::chrono::steady_clock::time_point expiry;
+        T user_data;
+    };
+
+    std::unordered_map<ConnectionID, SubData> subs;
+    std::shared_mutex _mutex;
+    const std::string description; // description of the sub for logging
+    const std::chrono::minutes sub_duration; // extended by re-subscribe
+
+public:
+
+    Subscription() = delete;
+    Subscription(std::string description, std::chrono::milliseconds sub_duration = 30min)
+        : description(std::move(description)), sub_duration(sub_duration)
```

Comment:
Style: Use curlies for member initialization: `description{std::move()}, sub_duration{sub_duration}`